designs/PackageVulnerability/PMUIVulnerability.md

```diff
@@ -0,0 +1,55 @@
+
+# PMUI Display vulnerability information in the package graph
+
+* Status: [In Review] / Reviewed / Implementing / Implemented
+* Author(s): [drewgillies](https://github.com/drewgillies)
+* Issues: 
+[1](https://github.com/NuGet/Home/issues/9850) Surface basic vulnerability metadata in PM UI view models
+[2](https://github.com/NuGet/Home/issues/8755) Display vulnerability info on Installed tab of PMUI
+
+## Problem Background
+
+The [Flag Vulnerable Packages spec](https://github.com/NuGet/Home/wiki/Flag-vulnerable-packages) covers a number of MVP initiatives of which this is one. The problem being solved is feature parity with existing package services (npm) couples with a need to service our customers' security needs.
+
+## Who are the customers
+
+All publishers of NuGet.org packages, all consumers of NuGet.org packages.
+
+## Goals
+
+When a customer browses their installed packages in the PMUI, the MVP is for them to see notifications on their installed tab for vulnerable packages. 
+
+## Non-Goals
+
+Notifications on the Browse tab
+Restore warnings
+
+## Solution
+
+The package vulnerability data is now available in registration metadata, alongside data for deprecations which is currently in use.
+
+The plan is the extend the design used for surfacing deprecation notifications in the past, to also include vulnerability information. This includes models, view models and views. The work is likely to be very similar in all areas. The strengths of this approach are a reasonably fast time to market, and low risk.
```

Comment:
As discussed briefly in the PR, but mostly in Teams, we've (very) recently started paying a lot more attention to the [NuGet SDK](https://github.com/NuGet/NuGet.Client/blob/dev/docs/nuget-sdk.md) than did in the past. I have no concerns from the customer experience point of view.

However, the existing deprecations implementation added a `GetDeprecationMetadataAsync` on `IPackageSearchMetadata` that's defined in `NuGet.Protocol.dll`, which is part of the NuGet SDK. Customers calling these methods in their own apps **always** get `null`, even though when they check the package on nuget.org in a web browser, it shows the information exists. This is just bad API design.

Considering the only implementation of `GetVersionsAsync` and `GetDeprecationMetadataAsync` that do not return null exist in some PM UI assembly, this is where the interface that defines the methods should exist, in my opinion. My suggestion, and I have no idea how feasible it actually is, is to add any new `GetVulnerabilityMetadataAsync` method to an interface/class/whatever in PM UI, not in `NuGet.Protocol`.

Anything that's not in the NuGet SDK, and therefore doesn't require changes to any `PublicAPI.*.txt` file to compile, is fine, because we're free to refactor as we wish. Any API listed in `PublicAPI.*.txt` might be used by customers (external to Microsoft, but also teams in Microsoft, such as the .NET CLI/SDK or VS for Mac), and therefore we should do a better job of designing APIs.

Reply:
I've created a prototype of the synchronous approach and pushed a commit here: https://github.com/NuGet/NuGet.Client/pull/3540. This makes me optimistic that we can take this approach. Not sure if any edge cases will fall out as we bash on it though, but I'll take this spec to a design review on the strength of it.